source/extensions/common/aws/utility.cc

```diff
@@ -58,32 +69,120 @@ Utility::canonicalizeHeaders(const Http::RequestHeaderMap& headers) {
   return out;
 }
 
-std::string
-Utility::createCanonicalRequest(absl::string_view method, absl::string_view path,
-                                const std::map<std::string, std::string>& canonical_headers,
-                                absl::string_view content_hash) {
+std::string Utility::createCanonicalRequest(
+    absl::string_view service_name, absl::string_view method, absl::string_view path,
+    const std::map<std::string, std::string>& canonical_headers, absl::string_view content_hash) {
   std::vector<absl::string_view> parts;
   parts.emplace_back(method);
   // don't include the query part of the path
-  const auto path_part = StringUtil::cropRight(path, "?");
-  parts.emplace_back(path_part.empty() ? "/" : path_part);
-  const auto query_part = StringUtil::cropLeft(path, "?");
+  const auto path_part = StringUtil::cropRight(path, QUERY_SPLITTER);
+  const auto canonicalized_path = path_part.empty()
+                                      ? std::string{PATH_SPLITTER}
+                                      : canonicalizePathString(path_part, service_name);
+  parts.emplace_back(canonicalized_path);
+  const auto query_part = StringUtil::cropLeft(path, QUERY_SPLITTER);
   // if query_part == path_part, then there is no query
-  parts.emplace_back(query_part == path_part ? "" : query_part);
+  const auto canonicalized_query =
+      query_part == path_part ? EMPTY_STRING : Utility::canonicalizeQueryString(query_part);
+  parts.emplace_back(absl::string_view(canonicalized_query));
   std::vector<std::string> formatted_headers;
   formatted_headers.reserve(canonical_headers.size());
   for (const auto& header : canonical_headers) {
     formatted_headers.emplace_back(fmt::format("{}:{}", header.first, header.second));
     parts.emplace_back(formatted_headers.back());
   }
   // need an extra blank space after the canonical headers
-  parts.emplace_back("");
+  parts.emplace_back(EMPTY_STRING);
   const auto signed_headers = Utility::joinCanonicalHeaderNames(canonical_headers);
   parts.emplace_back(signed_headers);
   parts.emplace_back(content_hash);
   return absl::StrJoin(parts, "\n");
 }
 
+std::string Utility::canonicalizePathString(absl::string_view path_string,
```

Comment:
Can you include the comment with the link to AWS reference, please ? https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html

Reply:
Thanks for the review :) 
I have already added the link in doc string in the header file. But sure I'll add it here too.